# **Machine Learning: Full Pipeline**

## Library Import, Constants

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split

In [2]:
RANDOM_STATE = 42

In [3]:
DATASET_PATH = "https://raw.githubusercontent.com/aiedu-courses/stepik_eda_and_dev_tools/main/datasets/online_shoppers_intention.csv"

## Loading and Preprocessing Data

Loading data:

In [4]:
df = pd.read_csv(DATASET_PATH)

Working with duplicates:

In [5]:
df.drop_duplicates(inplace = True)
df.reset_index(inplace = True, drop = True)

Working with missing values:

In [6]:
median_infdur = df['Informational_Duration'].median()
df['Informational_Duration'].fillna(median_infdur, inplace=True)

median_prdur = df['ProductRelated_Duration'].median()
df['ProductRelated_Duration'].fillna(median_prdur, inplace=True)

df = df.dropna(subset=['ExitRates'])

In [7]:
df['Month'] = df['Month'].replace('aug', 'Aug')

## Choosing a Target Metric

From the exploratory data analysis, we remember that the classes of the target variable are unbalanced

Let's choose the F1-score metric

F1-score is a harmonic mean between precision and recall. It works well for unbalanced classes because it takes into account both aspects of classification: accuracy in predicting a positive class and the ability of the model to detect all truly positive examples.

## NB

Let's build a Naive Bayesian Classifier model on numerical features with default parameters.

In [8]:
df

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12216               3                    145.0              0   
12217               0                      0.0              0   
12218               0                      0.0              0   
12219               4                     75.0              0   
12220               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12216                     0.0              53              1783.791667   
12217                     0.0               5               465.750000   
12218                     0.0               6               184.250000   
12219                     0.0              15               346.000000   
12220                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12216     0.007143   0.029031   12.241717         0.0   Dec                 4   
12217     0.000000   0.021333    0.000000         0.0   Nov                 3   
12218     0.083333   0.086667    0.000000         0.0   Nov                 3   
12219     0.000000   0.021053    0.000000         0.0   Nov                 2   
12220     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0            1       1            1  Returning_Visitor    False    False  
1            2       1            2  Returning_Visitor    False    False  
2            1       9            3  Returning_Visitor    False    False  
3            2       2            4  Returning_Visitor    False    False  
4            3       1            4  Returning_Visitor     True    False  
...        ...     ...          ...                ...      ...      ...  
12216        6       1            1  Returning_Visitor     True    False  
12217        2       1            8  Returning_Visitor     True    False  
12218        2       1           13  Returning_Visitor     True    False  
12219        2       3           11  Returning_Visitor    False    False  
12220        2       1            2        New_Visitor     True    False  

[12122 rows x 18 columns]

In [9]:
X = df[['Administrative',	'Administrative_Duration',	'Informational',
        'Informational_Duration',	'ProductRelated',	'ProductRelated_Duration',
        'BounceRates',	'ExitRates',	'PageValues',	'SpecialDay']]

y = df['Revenue'].astype(np.int)

<ipython-input-9-b2370b6eaf6d>:5: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  y = df['Revenue'].astype(np.int)


In [10]:
y.value_counts()

0    10236
1     1886
Name: Revenue, dtype: int64

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=RANDOM_STATE)

In [12]:
from sklearn.naive_bayes import GaussianNB

gnb = GaussianNB()

gnb.fit(X_train, y_train)

y_pred = gnb.predict(X_test)

In [13]:
from sklearn.metrics import f1_score

f1 = f1_score(y_test, y_pred)

print("F1-score:", f1)

F1-score: 0.5345528455284554


The value of F1-measure 0.53 can be interpreted as a relatively average result. Let's look at the confusion matrix to get more specific conclusions.


In [14]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred)

array([[2310,  235],
       [ 223,  263]])

In [15]:
y_test.value_counts()

0    2545
1     486
Name: Revenue, dtype: int64

We see that we have 2310 true positive observations, 235 false positive observations, 223 false negative observations, and 263 true negative observations.


That is, when predicting Revenue = 0, we have 9-10 times fewer incorrect responses than correct responses.

When predicting Revenue = 1, we got approximately the same number of correct and incorrect responses.

**We can see that the model is not very high quality.**

## KNN

Let's build a K-Nearest Neighbors model on numerical features with default parameters.

In [16]:
from sklearn.neighbors import KNeighborsClassifier

knn_cl = KNeighborsClassifier()

knn_cl.fit(X_train, y_train)

pred_knn = knn_cl.predict(X_test)

In [17]:
f1 = f1_score(y_test, pred_knn)

print("F1-score:", f1)

F1-score: 0.40169731258840174


The value of F1-measure 0.4 can be interpreted as a low result. The performance of the model in terms of the balance between accuracy and completeness is relatively low. **The model is worse.** Let's look at the confusion matrix to get more specific conclusions. The model has not improved.

In [18]:
confusion_matrix(y_test, pred_knn)

array([[2466,   79],
       [ 344,  142]])

We see that we have 2466 true positive observations, 79 false positive observations, 344 false negative observations, and 142 true negative observations.

That is, when predicting Revenue = 0, we have 31 times fewer incorrect responses than correct responses.

When predicting Revenue = 1, we got 2 times fewer correct responses than incorrect responses.

Perhaps such a model would be useful for ecommerce, but it turned out to be too "scrambled".

**We can see that the model is not very high quality.**

## KNN with GridSearchCV

Let's build a K-Nearest Neighbors model on numerical features with parameters matched with GridSearchCV

In [19]:
from sklearn.model_selection import GridSearchCV

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

model = KNeighborsClassifier()

params = {'n_neighbors' : np.arange(2, 20, 2),
          'weights' : ['uniform', 'distance'],
          'p' : [1, 2]}

gs = GridSearchCV(model, params, scoring='f1', cv=5, n_jobs=-1, verbose=2)
gs.fit(X_train, y_train)

Fitting 5 folds for each of 36 candidates, totalling 180 fits


GridSearchCV(cv=5, estimator=KNeighborsClassifier(), n_jobs=-1,
             param_grid={'n_neighbors': array([ 2,  4,  6,  8, 10, 12, 14, 16, 18]),
                         'p': [1, 2], 'weights': ['uniform', 'distance']},
             scoring='f1', verbose=2)

In [20]:
gs.best_score_, gs.best_params_

(0.41669366845958267, {'n_neighbors': 4, 'p': 2, 'weights': 'distance'})

In [21]:
pred = gs.best_estimator_.predict(X_test)

f1 = f1_score(y_test, pred)

print("F1-score:", f1)

F1-score: 0.4461343472750317


The value of F1-measure 0.4 can be interpreted as a low result. The performance of the model in terms of the balance between accuracy and completeness is relatively low. **The model is still not good.** Let's look at the confusion matrix to get more specific conclusions. The model has not improved.

In [22]:
confusion_matrix(y_test, pred)

array([[2418,  127],
       [ 310,  176]])

We see that we have 2418 true positive observations, 127 false positive observations, 310 false negative observations, and 176 true negative observations.

That is, when predicting Revenue = 0, we have 79 times fewer incorrect responses than correct responses.

When predicting Revenue = 1, we got 4 times fewer correct responses than incorrect responses.

We see that the model became better at predicting non-purchase but worse at predicting purchase. The model became even more "scared" than in the previous step.

## NB with GridSearchCV

Let's build a Naive Bayesian Classifier model on numerical features with parameters matched with GridSearchCV

In [23]:
from sklearn.model_selection import GridSearchCV

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=42)

model = GaussianNB()

params = {'priors': [None, [0.7, 0.3], [0.3, 0.7], [0.8, 0.2], [0.2, 0.8], [0.9, 0.1], [0.1, 0.9]],
          'var_smoothing': [1e-9, 1e-8, 1e-7, 1e-6]}

gs = GridSearchCV(model, params, scoring='f1', cv=5, n_jobs=-1, verbose=2)
gs.fit(X_train, y_train)


Fitting 5 folds for each of 28 candidates, totalling 140 fits


GridSearchCV(cv=5, estimator=GaussianNB(), n_jobs=-1,
             param_grid={'priors': [None, [0.7, 0.3], [0.3, 0.7], [0.8, 0.2],
                                    [0.2, 0.8], [0.9, 0.1], [0.1, 0.9]],
                         'var_smoothing': [1e-09, 1e-08, 1e-07, 1e-06]},
             scoring='f1', verbose=2)

In [24]:
gs.best_score_, gs.best_params_

(0.5373385426571803, {'priors': [0.3, 0.7], 'var_smoothing': 1e-06})

In [25]:
pred = gs.best_estimator_.predict(X_test)

f1 = f1_score(y_test, pred)

print("F1-score:", f1)

F1-score: 0.5694323144104804


The performance of the model in terms of the balance between accuracy and completeness is relatively high. The quality of the model, relative to this metric, does not differ from that of KNN.Let's look at the confusion matrix to get more specific conclusions.

The value of F1-measure 0.57 can be interpreted as a low result. The performance of the model in terms of the balance between accuracy and completeness is not bad. **The model has improved.** Let's look at the confusion matrix to get more specific conclusions.

In [26]:
confusion_matrix(y_test, pred)

array([[2212,  333],
       [ 160,  326]])

We see that we have 2212 true positive observations, 333 false positive observations, 160 false negative observations, and 326 true negative observations.

That is, when predicting Revenue = 0, we have 7 times fewer incorrect responses than correct responses.

When predicting Revenue = 1, we got 2 times fewer correct responses than incorrect responses.

We see that the model became better at predicting non-purchase but worse at predicting purchase. That said, the model isn't as "scuzzy" as past ones. It is a better predictor of buying than not buying.


## Keeping categorical attributes

### **Encoding**

In [27]:
df

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12216               3                    145.0              0   
12217               0                      0.0              0   
12218               0                      0.0              0   
12219               4                     75.0              0   
12220               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12216                     0.0              53              1783.791667   
12217                     0.0               5               465.750000   
12218                     0.0               6               184.250000   
12219                     0.0              15               346.000000   
12220                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12216     0.007143   0.029031   12.241717         0.0   Dec                 4   
12217     0.000000   0.021333    0.000000         0.0   Nov                 3   
12218     0.083333   0.086667    0.000000         0.0   Nov                 3   
12219     0.000000   0.021053    0.000000         0.0   Nov                 2   
12220     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0            1       1            1  Returning_Visitor    False    False  
1            2       1            2  Returning_Visitor    False    False  
2            1       9            3  Returning_Visitor    False    False  
3            2       2            4  Returning_Visitor    False    False  
4            3       1            4  Returning_Visitor     True    False  
...        ...     ...          ...                ...      ...      ...  
12216        6       1            1  Returning_Visitor     True    False  
12217        2       1            8  Returning_Visitor     True    False  
12218        2       1           13  Returning_Visitor     True    False  
12219        2       3           11  Returning_Visitor    False    False  
12220        2       1            2        New_Visitor     True    False  

[12122 rows x 18 columns]

In [28]:
{column: list(df[column].unique()) for column in df.columns if df.dtypes[column] == 'object'}

{'Month': ['Feb',
  'Aug',
  'Mar',
  'May',
  'Oct',
  'June',
  'Jul',
  'Nov',
  'Sep',
  'Dec'],
 'VisitorType': ['Returning_Visitor', 'New_Visitor', 'Other']}

We can encode the values of `Month` with ordinary integers in ascending order, so we will keep priorities. We will encode  `VisitorType` using One Hot Encoding.

Let's encode object columns

In [29]:
month_ordering = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'June', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
df['Month'] = df['Month'].apply(lambda x: month_ordering.index(x))

In [30]:
dummies = pd.get_dummies(df['VisitorType'], prefix='VT', drop_first=True)
df = pd.concat([df, dummies], axis=1)
df = df.drop('VisitorType', axis=1)

In [31]:
df

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12216               3                    145.0              0   
12217               0                      0.0              0   
12218               0                      0.0              0   
12219               4                     75.0              0   
12220               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12216                     0.0              53              1783.791667   
12217                     0.0               5               465.750000   
12218                     0.0               6               184.250000   
12219                     0.0              15               346.000000   
12220                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay  Month  \
0         0.200000   0.200000    0.000000         0.0      1   
1         0.000000   0.100000    0.000000         0.0      1   
2         0.200000   0.200000    0.000000         0.0      1   
3         0.050000   0.140000    0.000000         0.0      1   
4         0.020000   0.050000    0.000000         0.0      1   
...            ...        ...         ...         ...    ...   
12216     0.007143   0.029031   12.241717         0.0     11   
12217     0.000000   0.021333    0.000000         0.0     10   
12218     0.083333   0.086667    0.000000         0.0     10   
12219     0.000000   0.021053    0.000000         0.0     10   
12220     0.000000   0.066667    0.000000         0.0     10   

       OperatingSystems  Browser  Region  TrafficType  Weekend  Revenue  \
0                     1        1       1            1    False    False   
1                     2        2       1            2    False    False   
2                     4        1       9            3    False    False   
3                     3        2       2            4    False    False   
4                     3        3       1            4     True    False   
...                 ...      ...     ...          ...      ...      ...   
12216                 4        6       1            1     True    False   
12217                 3        2       1            8     True    False   
12218                 3        2       1           13     True    False   
12219                 2        2       3           11    False    False   
12220                 3        2       1            2     True    False   

       VT_Other  VT_Returning_Visitor  
0             0                     1  
1             0                     1  
2             0                     1  
3             0                     1  
4             0                     1  
...         ...                   ...  
12216         0                     1  
12217         0                     1  
12218         0                     1  
12219         0                     1  
12220         0                     0  

[12122 rows x 19 columns]

Let's encode boolean columns (`Weekend` and `Revenue`)

In [32]:
df['Weekend'] = df['Weekend'].astype(np.int)
df['Revenue'] = df['Revenue'].astype(np.int)
df

<ipython-input-32-5ebc9c763668>:1: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  df['Weekend'] = df['Weekend'].astype(np.int)
<ipython-input-32-5ebc9c763668>:2: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
Deprecated in NumPy 1

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12216               3                    145.0              0   
12217               0                      0.0              0   
12218               0                      0.0              0   
12219               4                     75.0              0   
12220               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12216                     0.0              53              1783.791667   
12217                     0.0               5               465.750000   
12218                     0.0               6               184.250000   
12219                     0.0              15               346.000000   
12220                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay  Month  \
0         0.200000   0.200000    0.000000         0.0      1   
1         0.000000   0.100000    0.000000         0.0      1   
2         0.200000   0.200000    0.000000         0.0      1   
3         0.050000   0.140000    0.000000         0.0      1   
4         0.020000   0.050000    0.000000         0.0      1   
...            ...        ...         ...         ...    ...   
12216     0.007143   0.029031   12.241717         0.0     11   
12217     0.000000   0.021333    0.000000         0.0     10   
12218     0.083333   0.086667    0.000000         0.0     10   
12219     0.000000   0.021053    0.000000         0.0     10   
12220     0.000000   0.066667    0.000000         0.0     10   

       OperatingSystems  Browser  Region  TrafficType  Weekend  Revenue  \
0                     1        1       1            1        0        0   
1                     2        2       1            2        0        0   
2                     4        1       9            3        0        0   
3                     3        2       2            4        0        0   
4                     3        3       1            4        1        0   
...                 ...      ...     ...          ...      ...      ...   
12216                 4        6       1            1        1        0   
12217                 3        2       1            8        1        0   
12218                 3        2       1           13        1        0   
12219                 2        2       3           11        0        0   
12220                 3        2       1            2        1        0   

       VT_Other  VT_Returning_Visitor  
0             0                     1  
1             0                     1  
2             0                     1  
3             0                     1  
4             0                     1  
...         ...                   ...  
12216         0                     1  
12217         0                     1  
12218         0                     1  
12219         0                     1  
12220         0                     0  

[12122 rows x 19 columns]

### **Splitting and Scaling**

In [33]:
X = df.drop('Revenue', axis=1)
y = df['Revenue']

In [34]:
Xtrain, Xtest, ytrain, ytest = train_test_split(X, y, test_size=0.3, random_state=RANDOM_STATE)

In [35]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(Xtrain)

Xtrain = scaler.transform(Xtrain)
Xtest = scaler.transform(Xtest)

In [36]:
Xtrain = pd.DataFrame(Xtrain, columns=X.columns)
Xtest = pd.DataFrame(Xtest, columns=X.columns)

In [37]:
Xtrain

Administrative  Administrative_Duration  Informational  \
0          -0.701679                -0.455635      -0.401593   
1          -0.401984                -0.367657       1.970228   
2          -0.701679                -0.455635      -0.401593   
3          -0.701679                -0.455635      -0.401593   
4          -0.701679                -0.455635      -0.401593   
...              ...                      ...            ...   
8480        0.197407                 0.143716      -0.401593   
8481       -0.401984                -0.400648      -0.401593   
8482       -0.701679                -0.455635      -0.401593   
8483       -0.701679                -0.455635      -0.401593   
8484       -0.701679                -0.455635      -0.401593   

      Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                  -0.239368       -0.113607                 0.071570   
1                   3.679180        0.177517                -0.221514   
2                  -0.239368       -0.539098                -0.400925   
3                  -0.239368        0.692585                 0.751241   
4                  -0.239368       -0.673463                -0.414090   
...                      ...             ...                      ...   
8480               -0.239368       -0.024031                 1.029271   
8481               -0.239368       -0.471915                -0.061428   
8482               -0.239368       -0.695857                -0.628454   
8483               -0.239368       -0.225579                -0.135986   
8484               -0.239368       -0.539098                -0.491538   

      BounceRates  ExitRates  PageValues  SpecialDay     Month  \
0       -0.449659  -0.417126   -0.326504   -0.308377  0.986859   
1       -0.209397  -0.390571   -0.326504   -0.308377  0.097953   
2       -0.449659  -0.629687   -0.326504   -0.308377  1.283161   
3        0.367681   0.554583   -0.326504   -0.308377 -0.198350   
4        1.789141   2.376536   -0.326504   -0.308377 -1.383558   
...           ...        ...         ...         ...       ...   
8480    -0.449659  -0.789108   -0.326504   -0.308377  1.283161   
8481    -0.001899  -0.028442   -0.326504   -0.308377 -0.790954   
8482     4.027942   3.469709   -0.326504   -0.308377  0.690557   
8483    -0.042604   0.157066   -0.326504   -0.308377 -1.383558   
8484    -0.449659  -0.278310   -0.326504   -0.308377 -0.198350   

      OperatingSystems   Browser    Region  TrafficType   Weekend  VT_Other  \
0            -0.133381 -0.209625 -0.893424    -0.768960 -0.548072 -0.085095   
1             0.987151 -0.209625  0.350742    -0.521496 -0.548072 -0.085095   
2             2.107682 -0.209625 -0.893424    -0.521496 -0.548072 -0.085095   
3             0.987151 -0.209625  2.009629     3.932849 -0.548072 -0.085095   
4            -0.133381  4.452677 -0.063980    -0.274033 -0.548072 -0.085095   
...                ...       ...       ...          ...       ...       ...   
8480          0.987151 -0.209625 -0.893424     1.458213 -0.548072 -0.085095   
8481         -0.133381 -0.209625 -0.893424    -0.521496  1.824577 -0.085095   
8482         -1.253912 -0.792413  2.009629    -0.768960  1.824577 -0.085095   
8483         -0.133381 -0.209625 -0.063980    -0.768960 -0.548072 -0.085095   
8484         -0.133381 -0.209625 -0.893424    -0.768960  1.824577 -0.085095   

      VT_Returning_Visitor  
0                 0.416827  
1                 0.416827  
2                -2.399078  
3                 0.416827  
4                 0.416827  
...                    ...  
8480              0.416827  
8481              0.416827  
8482              0.416827  
8483              0.416827  
8484              0.416827  

[8485 rows x 18 columns]

### **NB**

Let's focus on the NB model. Since it is known that it works better with a large number of attributes

In [38]:
model = GaussianNB()

params = {'priors': [None, [0.7, 0.3], [0.3, 0.7], [0.8, 0.2], [0.2, 0.8], [0.9, 0.1], [0.1, 0.9]],
          'var_smoothing': [1e-9, 1e-8, 1e-7, 1e-6]}

gs = GridSearchCV(model, params, scoring='f1', cv=5, n_jobs=-1, verbose=2)

gs.fit(Xtrain, ytrain)


Fitting 5 folds for each of 28 candidates, totalling 140 fits


GridSearchCV(cv=5, estimator=GaussianNB(), n_jobs=-1,
             param_grid={'priors': [None, [0.7, 0.3], [0.3, 0.7], [0.8, 0.2],
                                    [0.2, 0.8], [0.9, 0.1], [0.1, 0.9]],
                         'var_smoothing': [1e-09, 1e-08, 1e-07, 1e-06]},
             scoring='f1', verbose=2)

In [39]:
gs.best_score_, gs.best_params_

(0.5092227884979893, {'priors': [0.9, 0.1], 'var_smoothing': 1e-09})

In [40]:
pred = gs.best_estimator_.predict(Xtest)

f1_score(ytest, pred)

0.5428571428571428

The performance of the model in terms of the balance between accuracy and completeness is **relatively high**.

The value of F1-measure 0.54 indicates that the model demonstrates a good balance between accuracy and completeness in predictions.

Let's look at the confusion matrix to get more specific conclusions.

In [41]:
confusion_matrix(ytest, pred)

array([[2668,  394],
       [ 214,  361]])

We see that we have 2668 true positive observations, 394 false positive observations, 214 false negative observations, and 361 true negative observations.

That is, when predicting Revenue = 0, we have 7 times fewer incorrect responses than correct responses.

When predicting Revenue = 1, we got 1.7 times fewer correct responses than incorrect responses.

We see that the model became better at predicting non-purchase and purchase cases. The model became less "scared" than in the previous steps.

**We've made the model better**

## Explainer Dashboard

###**Building and Saving**

In [42]:
!pip install explainerdashboard -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 287.2/287.2 kB 3.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 220.6/220.6 kB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.4/10.4 MB 21.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.8/91.8 kB 8.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.9/547.9 kB 37.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.7/57.7 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.6/233.6 kB 22.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 136.5/136.5 kB 14.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 40.0 MB/s eta 0:00:00


In [43]:
from explainerdashboard import ClassifierExplainer, ExplainerDashboard

Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)


In [44]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [45]:
explainer = ClassifierExplainer(gs.best_estimator_, Xtest.iloc[:10], ytest.iloc[:10])

Note: shap values for shap='kernel' normally get calculated against X_background, but paramater X_background=None, so setting X_background=shap.sample(X, 50)...
Generating self.shap_explainer = shap.KernelExplainer(model, X, link='identity')


In [46]:
db = ExplainerDashboard(explainer)

Building ExplainerDashboard..
Detected google colab environment, setting mode='external'
For this type of model and model_output interactions don't work, so setting shap_interaction=False...
The explainer object has no decision_trees property. so setting decision_trees=False...
Generating layout...
Calculating shap values...


/usr/local/lib/python3.10/dist-packages/dash/dash.py:516: UserWarning: JupyterDash is deprecated, use Dash instead.
See https://dash.plotly.com/dash-in-jupyter for more details.


  0%|          | 0/10 [00:00<?, ?it/s]

Calculating prediction probabilities...
Calculating metrics...
Calculating confusion matrices...
Calculating classification_dfs...
Calculating roc auc curves...
Calculating pr auc curves...
Calculating liftcurve_dfs...
Calculating dependencies...
Calculating permutation importances (if slow, try setting n_jobs parameter)...
Calculating predictions...
Calculating pred_percentiles...
Reminder: you can store the explainer (including calculated dependencies) with explainer.dump('explainer.joblib') and reload with e.g. ClassifierExplainer.from_file('explainer.joblib')
Registering callbacks...


In [47]:
db.run()

Starting ExplainerDashboard on http://172.28.0.12:8050
You can terminate the dashboard with ExplainerDashboard.terminate(8050)
Dash app running on:


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [48]:
db.save_html('dashboard.html')

### **Analysis**

#### **Feature Importances**

From the plot in the dashboard, we can see that the most important feature in the data is the `PageValues` feature.

This feature stands out significantly among the others

The average value of the page plays the most important role in the model

#### **Metrics**

Let's look at the metrics that were presented in the dashboard

|   Metric         |   Score   |
|------------------|-----------|
|   accuracy       |   0.9     |
|   precision      |   0.5     |
|   recall         |   1.0     |
|   f1             |   0.667   |
|   roc_auc_score  |   1.0     |
|   pr_auc_score   |   1.0     |
|   log_loss       |   0.581   |

Precision is 0.5, which means that 50% of the objects that the model predicted
as class 1 were correctly predicted.

Log loss is 0.581, which means that the model predicts class probabilities quite accurately.

**Some of the metrics from the dashboard raise questions. The confusion matrix we built earlier suggests that not all classes were correctly predicted, but the metrics say otherwise:**

* Recall is 1, which means that the model correctly predicted all objects of class 1.

* Accuracy is 0.9, which means that the model correctly predicted classes for 90% of all objects.

* F1-score is 0.667, which means that the model has achieved a balance between Precision and Recall.

* The ROC AUC is equal to 1, which means that the model perfectly discriminates between classes.

* PR AUC is 1, which means that the model perfectly distinguishes between classes.


#### **Individual Predictions**

Let's look at how variables affect some individual predictions.

Let's look at the prediction with **index 3**. We can see that `PageValues` had a positive impact: +66.15%. At the same time, the other attributes did not have much impact. Because of this influence of `PageValues` the initial prediction became equal to 1.

Let's look at the prediction with **index 4**. We can see that `PageValues` had a negative impact: -18.47%. At the same time, the `Weekend` feature had a positive impact: +0.9%. Due to this influence of `PageValues` the initial prediction became equal to 0 (96.4%).

Let's look at the prediction with **index 5**. We see that `PageValues` had a positive impact: +54.06%. Also, `ExitRates` and `BounceRates` give postive impact (11-12%). At the same time, the `ProductRelated_Duration` feature had a negative impact: -0.8%. Due to this influence of variables, the prediction became equal to 1 (99%).